modules/libScePad/interfaces/ixip.cpp

```diff
@@ -0,0 +1,163 @@
+#include "core/timer/timer.h"
+#include "core/videoout/videoout.h"
+#include "ikbd.h"
+#include "logging.h"
+
+#include <bitset>
+#include <math.h>
+
+LOG_DEFINE_MODULE(libScePad_xip);
+
+static bool is_XIP_inited = false;
+
+class XIPController: public IController {
+  public:
+  XIPController(ControllerConfig* cfg, uint32_t userid): IController(ControllerType::SDL, cfg, userid) {
+    init();
+    reconnect();
+  }
+
+  virtual ~XIPController() = default;
+
+  // ### Interface
+  bool reconnect() final;
+  void close() final;
+  bool readPadData(ScePadData& data) final;
+  bool setMotion(bool state) final;
+  bool setRumble(const ScePadVibrationParam* pParam) final;
+  bool setLED(const ScePadColor* pParam) final;
+  bool resetLED() final;
+
+  uint32_t getButtons();
+
+  static void init();
+};
+
+std::unique_ptr<IController> createController_xinput(ControllerConfig* cfg, uint32_t userid) {
+  return std::make_unique<XIPController>(cfg, userid);
+}
+
+void XIPController::init() {
+  LOG_USE_MODULE(libScePad_xip);
+
+  if (is_XIP_inited == false) {
+    // todo: load xinput
+    LOG_CRIT(L"XInput support is not ready yet");
+    is_XIP_inited = true;
+  }
+}
+
+void XIPController::close() {
+  if (m_state == ControllerState::Disconnected || m_state == ControllerState::Closed) return;
+  m_state = ControllerState::Closed;
+}
+
+bool XIPController::reconnect() {
+  return false;
+}
+
+uint32_t XIPController::getButtons() {
+  std::bitset<32> bits;
```

Comment:
do you wan't to change the values later? bits is default = 0

Reply:
Yep, just a stub to change these values later, but... Actually iirc XInput stores pressed buttons in `int` value too, so I'll remove this bc we just need to convert bitsets between XInput and libScePad